### PR TITLE
[MRG] Cleaning and adding 'group_reid' function in 'noise_std' module

### PR DESCRIPTION
**Pull request by ja-che**

### CODE REVIEW

**Comment by bthirion** [comment]
Please let us know when this is ready for review.
Best,

**Comment by ja-che** [comment]
@bthirion it is ready for review (I thought that "[MRG]" was a kind way of asking for review).

File: hidimstat/noise_std.py [hunk 1]
```diff
@@ -1,66 +1,259 @@
...
+    error = clf_lasso_cv.predict(X) - y
+    coef_max = np.max(np.abs(beta_hat))
+
+    if coef_max == 0:
+        support = 0
+    else:
+        support = np.sum(np.abs(beta_hat) > tol * coef_max)
+
+    # avoid dividing by 0
+    if support >= n_samples:
```
Comment (bthirion) [inline hidimstat/noise_std.py:72]:
```suggestion
    support = np.minimum(support, n_samples - 1)
```
But  is it ever needed ?

File: hidimstat/noise_std.py [hunk 2]
```diff
@@ -1,66 +1,259 @@
...
+            support = np.sum(np.sum(np.abs(Beta_hat), axis=0) > tol * row_max)
+
+    else:
+
+        Beta_hat = np.zeros((n_features, n_targets))
+        Error = np.copy(Y)
+        support = 0
+
+    # avoid dividing by 0
+    if support >= n_samples:
```
Comment (bthirion) [inline hidimstat/noise_std.py:179]: same as above

File: hidimstat/noise_std.py [hunk 3]
```diff
@@ -1,66 +1,259 @@
...
+        Seed passed in the KFold object which is used to cross-validate
+        LassoCV. This seed controls the partitioning randomness.
+
+    Returns
+    -------
+    sigma_hat : float
+        Estimated noise standard deviation.
+
+    beta_hat : array, shape (n_features,)
+        Estimated parameter vector.
```
Comment (bthirion) [inline hidimstat/noise_std.py:45]: Do you provide somewhere a reference to Reid's paper ?

File: hidimstat/scenario.py [hunk 4]
```diff
@@ -182,3 +182,58 @@ def multivariate_simulation(n_samples=100,
...
+
+    X = rng.standard_normal((n_samples, n_features))
+
+    Beta = np.zeros((n_features, n_targets))
+    Beta[0:support_size, :] = 1.0
+
+    E = np.zeros((n_samples, n_targets))
+    E[:, 0] = rng.standard_normal(n_samples)
+
+    for i in np.arange(1, n_targets):
```
Comment (bthirion) [inline hidimstat/scenario.py:231]: This code seems quite inefficient: it seems like a manual filtering operation.

**Review by bthirion — COMMENTED** [review]
Thx ! 
Please also beware untested lines.

File: hidimstat/noise_std.py [hunk 5]
```diff
@@ -1,66 +1,259 @@
 import numpy as np
 from numpy.linalg import norm
-from sklearn.linear_model import LassoCV, LassoLarsCV
+from scipy.linalg import toeplitz, solve
+from sklearn.linear_model import LassoCV, MultiTaskLassoCV
+from sklearn.model_selection import KFold
 
 
-def reid(X, y, method="lars", tol=1e-4, max_iter=1e+3, n_jobs=1):
+def reid(X, y, eps=1e-2, tol=1e-4, max_iter=1e+4, n_jobs=1, seed=0):
```
Comment (agramfort) [inline hidimstat/noise_std.py:8]:
```suggestion
def reid(X, y, eps=1e-2, tol=1e-4, max_iter=1e4, n_jobs=1, seed=0):
```

File: hidimstat/noise_std.py (hunk 3 above)
Comment (agramfort) [reply hidimstat/noise_std.py:45]: you need to add a References section and use rst footnotes as done in sklearn for example

File: hidimstat/noise_std.py [hunk 6]
```diff
@@ -1,66 +1,259 @@
...
+        LassoCV(eps=eps, normalize=False, fit_intercept=False,
+                cv=cv, tol=tol, max_iter=max_iter, n_jobs=n_jobs)
+
+    clf_lasso_cv.fit(X, y)
+    beta_hat = clf_lasso_cv.coef_
+    error = clf_lasso_cv.predict(X) - y
+    coef_max = np.max(np.abs(beta_hat))
+
+    if coef_max == 0:
+        support = 0
```
Comment (agramfort) [inline hidimstat/noise_std.py:67]: are you sure this if branch is needed?

File: hidimstat/noise_std.py [hunk 7]
```diff
@@ -1,66 +1,259 @@
...
+        support = n_samples - 1
+
+    sigma_hat = norm(error) / np.sqrt(n_samples - support)
+
+    return sigma_hat, beta_hat
+
+
+def group_reid(X, Y, fit_Y=True, stationary=True, method='simple', order=1,
+               eps=1e-2, tol=1e-4, max_iter=1e+4, n_jobs=1, seed=0):
+
```
Comment (agramfort) [inline hidimstat/noise_std.py:82]:
```suggestion
               eps=1e-2, tol=1e-4, max_iter=1e4, n_jobs=1, seed=0):
```

File: hidimstat/noise_std.py [hunk 8]
```diff
@@ -1,66 +1,259 @@
...
+        Target.
+
+    fit_Y : bool, optional (default=True)
+        If True, Y will be regressed against X by MultiTaskLassoCV
+        and the covariance matrix is estimated on the residuals.
+        Otherwise, covariance matrix is estimated directly on Y.
+
+    stationary : bool, optional (default=True)
+        If True, noise is considered to have the same magnitude at each
+        time point. Otherwise, magnitude of the noise is not constant.
```
Comment (agramfort) [inline hidimstat/noise_std.py:100]:
```suggestion
        If True, noise is considered to have the same magnitude for each
        target. Otherwise, magnitude of the noise is not constant.
```

File: hidimstat/noise_std.py [hunk 9]
```diff
@@ -1,66 +1,259 @@
...
+
+    method : bool, optional (default='simple')
+        If 'simple', the correlation matrix is estimated by taking the
+        median of the correlation between two consecutive time points
```
Comment (agramfort) [inline hidimstat/noise_std.py:104]: you talk about time points but above Y is (n_samples, n_targets)
either use the word target or replace n_targets by n_times

File: hidimstat/noise_std.py [hunk 10]
```diff
@@ -1,66 +1,259 @@
...
+
+    max_iter : int, optional (default=1e+4)
+        The maximum number of iterations.
+
+    n_jobs : int or None, optional (default=1)
+        Number of CPUs to use during the cross validation.
+
+    seed: int, optional (default=0)
+        Seed passed in the KFold object which is used to cross-validate
+        LassoCV. This seed controls the partitioning randomness.
```
Comment (agramfort) [inline hidimstat/noise_std.py:128]:
```suggestion
        LassoCV. This seed controls also the partitioning randomness.
```

File: hidimstat/noise_std.py [hunk 11]
```diff
@@ -1,66 +1,259 @@
...
+    """
+
+    X = np.asarray(X)
+    n_samples, n_features = X.shape
+    n_targets = Y.shape[1]
+
+    if method == 'simple':
+        print('Group reid: simple cov estimation')
+    else:
+        print('Group reid: ' + method + str(order) + ' cov estimation')
```
Comment (agramfort) [inline hidimstat/noise_std.py:146]: use an f-string

File: hidimstat/noise_std.py [hunk 12]
```diff
@@ -1,66 +1,259 @@
...
+
+    if (max_iter // 5) <= n_features:
+        max_iter = n_features * 5
+        print("'max_iter' has been increased to {}".format(max_iter))
```
Comment (agramfort) [inline hidimstat/noise_std.py:150]:
```suggestion
        print(f"'max_iter' has been increased to {max_iter}")
```

File: hidimstat/noise_std.py [hunk 13]
```diff
@@ -1,66 +1,259 @@
...
+
+    if fit_Y:
+
+        clf_mtlcv = \
+            MultiTaskLassoCV(eps=eps, normalize=False, fit_intercept=False,
+                             cv=cv, tol=tol, max_iter=max_iter,
+                             n_jobs=n_jobs)
+
+        clf_mtlcv.fit(X, Y)
+        Beta_hat = clf_mtlcv.coef_
```
Comment (agramfort) [inline hidimstat/noise_std.py:162]: please not capital letters in var names. The only exception is for X and Y
```suggestion
        beta_hat = clf_mtlcv.coef_
```

File: hidimstat/noise_std.py [hunk 14]
```diff
@@ -1,66 +1,259 @@
...
+        Error = clf_mtlcv.predict(X) - Y
```
Comment (agramfort) [inline hidimstat/noise_std.py:163]:
```suggestion
        residual = clf_mtlcv.predict(X) - Y
```

File: hidimstat/noise_std.py [hunk 15]
```diff
@@ -1,66 +1,259 @@
...
+        coef_max = np.max(np.abs(Beta_hat))
+        row_max = np.max(np.sum(np.abs(Beta_hat), axis=0))
+
+        if coef_max == 0:
+            support = 0
```
Comment (agramfort) [inline hidimstat/noise_std.py:168]: needed?

File: hidimstat/noise_std.py [hunk 16]
```diff
@@ -1,66 +1,259 @@
...
+
+    # avoid dividing by 0
+    if support >= n_samples:
+        support = n_samples - 1
+
+    sigma_hat_raw = norm(Error, axis=0) / np.sqrt(n_samples - support)
+
+    if stationary:
+        sigma_hat = np.median(sigma_hat_raw) * np.ones(n_targets)
+        Corr_emp = np.corrcoef(Error.T)
```
Comment (agramfort) [inline hidimstat/noise_std.py:186]:
```suggestion
        corr_emp = np.corrcoef(Error.T)
```

File: hidimstat/scenario.py [hunk 17]
```diff
@@ -182,3 +182,58 @@ def multivariate_simulation(n_samples=100,
...
+                                     sigma=1.0, rho=0.0, seed=0):
+    """Generate 1D temporal data with constant design matrix
+
+    Parameters
+    -----------
+    n_samples : int
+        Number of samples
+    n_features : int
+        Number of features
+    n_targets : int
```
Comment (agramfort) [inline hidimstat/scenario.py:198]:
```suggestion
    n_times : int
```

File: hidimstat/noise_std.py (hunk 16 above)
Comment (ja-che) [reply hidimstat/noise_std.py:186]: `Corr_emp` is a matrix

File: hidimstat/scenario.py (hunk 4 above)
Comment (ja-che) [reply hidimstat/scenario.py:231]: The advantage of this method is that it controls the level of autocorrelation precisely while with Gaussian filtering it is not possible. And, I did not find any shortcut to do that (neither in numpy or scipy).

File: hidimstat/noise_std.py (hunk 16 above)
Comment (ja-che) [reply hidimstat/noise_std.py:186]: But ok, if I see that it does not matter, capital letters only for `X` and `Y`.

**Review by agramfort — APPROVED** [review]
